File: Medium/5. Longest Palindromic Substring/LongestPalindromicSubstring.py
'''
LeetCode Medium: 5. Longest Palindromic Substring
https://leetcode.com/problems/longest-palindromic-substring/description/

Check LongestPalindromicSubstring.cpp for brief solution explanation.
'''
class Solution:
    def longestPalindrome(self, s: str) -> str:
        N = len(s)
        dp = [[False for i in range(N)] for j in range(N)]
        st, ed = 0, 0
        for length in range(1, N+1):
            for i in range(N-length+1):
                j = i+length-1
                if length<=2:
                    dp[i][j] = s[i]==s[j]
                else:
                    dp[i][j] = dp[i+1][j-1] & (s[i]==s[j])
                # Keep track of best
                if dp[i][j] and ed-st+1<=length:
                    st, ed = i, j
        return s[st:ed+1]

File: Medium/5. Longest Palindromic Substring/LongestPalindromicSubstring.java
/*
LeetCode Medium: 5. Longest Palindromic Substring
https://leetcode.com/problems/longest-palindromic-substring/description/

Check LongestPalindromicSubstring.cpp for brief solution explanation.
*/
class Solution {
    public String longestPalindrome(String s) {
        int N = s.length();
        boolean[][]dp = new boolean[N][N];
        int st = 0, ed = 0;
        for(int len = 1;len<=N;len++){
            for(int i = 0;i<N-len+1;i++){
                int j = i+len-1;
                if(len<=2){
                    dp[i][j] = (s.charAt(i)==s.charAt(j));
                }else{
                    dp[i][j] = dp[i+1][j-1] & (s.charAt(i)==s.charAt(j));
                }
                //Check for longest palindrome
                if(dp[i][j] && ed-st+1<=len){
                    st = i;
                    ed = j;
                }
            }
        }
        return s.substring(st, ed+1);
    }
}